lib/solarwinds_apm/sampling/token_bucket.rb

```diff
@@ -13,123 +13,145 @@ class TokenBucket
+          SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] Updated interval: #{@interval}ms" }
+        end
       end
 
-      self.rate = settings.rate if settings.rate
+      if settings[:interval] && running
+        stop
+        start
+      end
 
-      return unless settings.interval
+      start unless running
+    end
 
-      self.interval = settings.interval
-      return unless running
+    def tb_to_hash(settings)
+      { capacity: settings.capacity,
+        rate: settings.rate,
+        interval: settings.interval }
+    end
 
-      stop
-      start
+    def capacity
+      @lock.synchronize { @capacity }
     end
 
     def capacity=(capacity)
-      @capacity = [0, capacity].max
+      @lock.synchronize { @capacity = [0, capacity].max }
+    end
+
+    def rate
+      @lock.synchronize { @rate }
     end
 
     def rate=(rate)
-      @rate = [0, rate].max
+      @lock.synchronize { @rate = [0, rate].max }
+    end
+
+    def interval
+      @lock.synchronize { @interval }
     end
 
-    # self.interval= sets the @interval and @sleep_interval
-    # @sleep_interval is used in the timer thread to sleep between replenishing the bucket
     def interval=(interval)
-      @interval = interval.clamp(0, MAX_INTERVAL)
-      @sleep_interval = @interval / 1000.0
+      @lock.synchronize { @interval = interval.clamp(0, MAX_INTERVAL) }
+    end
+
+    def tokens
+      @lock.synchronize { @tokens }
     end
 
     def tokens=(tokens)
-      @tokens = tokens.clamp(0, @capacity)
+      @lock.synchronize { @tokens = tokens.clamp(0, @capacity) }
     end
 
     # Attempts to consume tokens from the bucket
     # @param n [Integer] Number of tokens to consume
     # @return [Boolean] Whether there were enough tokens
-    # TODO: we need to include thread-safety here since sampler is shared across threads
-    # and we may have multiple threads trying to consume tokens at the same time
     def consume(token = 1)
-      if @tokens >= token
-        self.tokens = @tokens - token
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] #{@type} Consumed #{token} from total #{@tokens} (#{(@tokens.to_f / @capacity * 100).round(1)}% remaining)" }
-        true
-      else
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] #{@type} Token consumption failed: requested=#{token}, available=#{@tokens}, capacity=#{@capacity}" }
-        false
+      @lock.synchronize do
+        if @tokens >= token
+          @tokens -= token
+          SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] Consumed #{token} tokens, remaining: #{@tokens}/#{@capacity} (#{(@tokens.to_f / @capacity * 100).round(1)}%)" }
+          true
+        else
+          SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] Token consumption failed: requested=#{token}, available=#{@tokens}, capacity=#{@capacity}" }
+          false
+        end
       end
     end
 
     # Starts replenishing the bucket
     def start
```

Comment:
The token bucket replenishment approach is different from JS/Python--it seems a timer thread on interval adds the token here, while the other two implementations "replenish during consume", meaning, at the time a token consume call is made, the interval between last consume and now is calculated, then the number of tokens that accrued during the interval is added.

Was there a reason you went with this different approach?

Reply:
Ah my bad, i just assumed the Python implementation is based on JS.

Question for @tammy-baylis-swi or @jerrytfleung now... do you remember the reason for Python implementation to not use a background thread to replenish tokens?

And @xuan-cao-swi how does the "reinit thread for forked child process" work in Ruby--not seeing anything obvious in this PR that corresponds to the Python register_at_fork.